go/pkg/io/impath/impath.go

```diff
@@ -141,22 +145,78 @@ func MustRel(elements ...string) Relative {
 	return p
 }
 
-// Descendant returns a relative path to the specified base path such that
-// when joined together with the base using filepath.Join(base, path), the result
-// is lexically equivalent to the specified target path.
+// Descendant returns a relative path to the base such that when joined together with base using
+// filepath.Join(base, path), the result is lexically equivalent to target.
 //
-// The returned error is nil, ErrNotRelative, or ErrNotDescendant.
-// ErrNotRelative indicates that the target cannot be made relative to base.
-// ErrNotDescendant indicates the target is not a descendant of base, even though it is relative.
+// The returned error is nil or ErrNotDescendant.
 func Descendant(base Absolute, target Absolute) (Relative, error) {
-	p, err := filepath.Rel(base.String(), target.String())
-	if err != nil {
+	b := base.String()
+	// If not the root itself (unix), add a separator.
+	if !os.IsPathSeparator(b[len(b)-1]) {
+		b += string(os.PathSeparator)
+	}
+	rel := strings.TrimPrefix(target.String(), b)
+	if rel == target.String() {
 		// On unix, this should never happen since all absolute paths share the same root.
 		// On Windows, it's possible for two absolute paths to have different roots (different drive letters).
-		return zeroRel, errors.Join(ErrNotRelative, err)
-	}
-	if strings.HasPrefix(p, "..") {
 		return zeroRel, errors.Join(ErrNotDescendant, fmt.Errorf("target %q is not a descendant of base %q", target, base))
 	}
-	return Relative{path: p}, nil
+	return Relative{path: rel}, nil
+}
+
+// fastDir assumes all paths are clean and avoids calling filepath.Clean.
+// Code is taken from filepath.Dir of go1.19.
+func fastDir(path string) string {
+	vol := filepath.VolumeName(path)
+	i := len(path) - 1
+	for i >= len(vol) && !os.IsPathSeparator(path[i]) {
+		i--
+	}
+	dir := path[len(vol) : i+1]
+	if dir == "." && len(vol) > 2 {
+		// must be UNC
+		return vol
+	}
+	if os.IsPathSeparator(dir[len(dir)-1]) {
+		dir = dir[:len(dir)-1]
+	}
+	return vol + dir
+}
+
+// clean returns a clean path from provided parts.
+func clean(base string, elements []Relative) string {
+	paths := make([]string, 0, len(elements)+1)
+	if base != "" {
+		paths = append(paths, base)
+	}
+	for _, p := range elements {
+		if p.String() == "" {
+			continue
+		}
+		paths = append(paths, p.String())
+	}
+	path := strings.Join(paths, string(os.PathSeparator))
+	if dirty(path) {
+		path = filepath.Clean(path)
+	}
+	return path
+}
+
+// dirty return true if path contains consecutive separators or a dot elements (e.g. . or ..).
+func dirty(path string) bool {
+	dotsElm := true
+	lastRune := '\000'
```

Comment:
optional: maybe `prevRune` instead. One could think that last refers to the last character in path, not the one previously encountered.